src/game/requester.c

```diff
@@ -174,34 +185,61 @@ void Requester_SetHeading(REQUEST_INFO *req, const char *string)
     Text_Remove(req->heading);
     req->heading = NULL;
 
-    if (string) {
-        req->heading_text = string; // unsafe
-    } else {
-        req->heading_text = NULL;
-    }
+    const size_t out_size = snprintf(NULL, 0, "%s", string) + 1;
+    snprintf(req->heading_text, out_size, "%s", string);
 }
 
 void Requester_ChangeItem(
-    REQUEST_INFO *req, int32_t idx, const char *string, uint16_t flag)
+    REQUEST_INFO *req, int32_t idx, bool blocked, const char *fmt, ...)
 {
-    if (string) {
-        strcpy(&req->item_texts[idx * req->item_text_len], string);
-        req->item_flags[idx] = flag;
-    } else {
-        req->item_flags[idx] = 0;
+    if (idx < 0 || idx >= req->max_items || !fmt) {
+        return;
+    }
+
+    if (req->item_texts[idx]) {
+        Memory_FreePointer(&req->item_texts[idx]);
+    }
+
+    va_list va;
+    va_start(va, fmt);
+    const size_t out_size = vsnprintf(NULL, 0, fmt, va) + 1;
+    req->item_texts[idx] = Memory_Alloc(sizeof(char) * out_size);
+    va_end(va);
+    Memory_FreePointer(&req->item_texts[idx]);
+
+    if (req->item_texts[idx] != NULL) {
+        va_start(va, fmt);
+        vsnprintf(req->item_texts[idx], out_size, fmt, va);
+        req->item_blocked[idx] = blocked;
+        va_end(va);
```

Comment:
I think we can simplify this function a bit:
1. `Memory_Alloc` is guaranteed to crash if the game fails to allocate the memory, so the NULL check can be removed.
2. After applying 1., since we no longer care about special scenarios, we can forgo the middle `va_end` / `va_start` calls and do `va_start()`, `vsnprintf(NULL, 0, …)`, `vsnprintf(buffer, out_size, …)`, `va_end()` in succession.